File: spotifyarch/src/main/java/com/timothy/spotifyarchitecture/room/SpotifyModule.java
package com.timothy.spotifyarchitecture.room;

import android.app.Application;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;

import com.timothy.spotifyarchitecture.SpotifyRepository;
import com.timothy.spotifyarchitecture.retrofit.SpotifyCreator;
import com.timothy.spotifyarchitecture.retrofit.SpotifyService;
import com.timothy.spotifyarchitecture.retrofit.SpotifyServiceAuth;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * Created by tim on 5/31/17.
 */
public interface SpotifyModule {

    SpotifyCreator.ApiAuthenticator provideApiAuthenticator();

    SpotifyService provideSpotifyService(SpotifyCreator.ApiAuthenticator apiAuthenticator);

    SpotifyServiceAuth provideSpotifyServiceAuth();

    SpotifyRepository provideSpotifyRepository(SpotifyDao spotifyDao, SpotifyService spotifyService, SpotifyServiceAuth spotifyServiceAuth, SpotifyCreator.ApiAuthenticator apiAuthenticator);

}

File: spotifyarch/src/main/java/com/timothy/spotifyarchitecture/room/SpotifyModule.kt
package com.timothy.spotifyarchitecture.room

import com.timothy.spotifyarchitecture.SpotifyRepository
import com.timothy.spotifyarchitecture.retrofit.SpotifyCreator
import com.timothy.spotifyarchitecture.retrofit.SpotifyService
import com.timothy.spotifyarchitecture.retrofit.SpotifyServiceAuth

/**
 * Dagger module interface to provide necessary objects upon startup.
 */
interface SpotifyModule {
    
    fun provideSpotifyDatabase(): SpotifyDatabase
    
    fun provideSpotifyDao(spotifyDatabase: SpotifyDatabase): SpotifyDao
    
    fun provideApiAuthenticator(): SpotifyCreator.ApiAuthenticator

    fun provideSpotifyService(apiAuthenticator: SpotifyCreator.ApiAuthenticator): SpotifyService

    fun provideSpotifyServiceAuth(): SpotifyServiceAuth

    fun provideSpotifyRepository(spotifyDao: SpotifyDao, spotifyService: SpotifyService, spotifyServiceAuth: SpotifyServiceAuth, apiAuthenticator: SpotifyCreator.ApiAuthenticator): SpotifyRepository

}
